### PR TITLE
Validation with Serpent

### PR DESCRIPTION
**Pull request by amandalund**
There are a few changes introduced in this PR. The main addition is the ability to compare OpenMC results to Serpent 2 results in each set of validation scripts. This can be done using the option `-c serpent` (default is `mcnp`). In addition, when creating a cross section library on the fly, the path to an XSDIR directory file should be specified rather than the path to the directory containing the ACE library. This will be used to locate the ACE table and create an HDF5 library for use with OpenMC (and an XSDATA directory file if the comparison is with Serpent). For photon transport calculations with Serpent, a path to the directory containing the additional Serpent photon data is also required. Finally, for neutron transport validation there is now an option to include thermal scattering data by passing in the ZAID of the S(alpha, beta) table, e.g., `-t grph.10t`.

### CODE REVIEW

**Comment by amandalund** [comment]
I also took a look at some of the cases where we were still seeing differences between OpenMC and MCNP with photon calculations (namely high energies, high-Z elements, and compounds), and as expected we see much better agreement with Serpent:

U-1 0e+02MeV_mcnp
U-1 0e+02MeV_serpent

H20-1 0e+02MeV_mcnp
H20-1 0e+02MeV_serpent

File: neutron-physics/model.py [hunk 1]
```diff
@@ -108,80 +50,276 @@ class Model(object):
...
+    def particles(self, particles):
+        if particles % self._batches != 0:
+            msg = (f'Number of particles {particles} must be divisible by '
+                   f'the number of batches {self._batches}.')
+            raise ValueError(msg)
+        self._particles = particles
+
+    @code.setter
+    def code(self, code):
+        if code not in ['mcnp', 'serpent']:
```
Comment (paulromano) [inline neutron-physics/model.py]: `if code not in ('mcnp', 'serpent'):`

File: neutron-physics/model.py [hunk 2]
```diff
@@ -108,80 +50,276 @@ class Model(object):
...
+            msg = (f'Unsupported code {code}: code must be either "mcnp" or '
+                   f'"serpent".')
```
Comment (paulromano) [inline neutron-physics/model.py]: Nothing is formatted on the second literal here so you can remove the f-prefix.

File: neutron-physics/model.py [hunk 3]
```diff
@@ -108,80 +50,276 @@ class Model(object):
...
+            raise ValueError(msg)
+        executable = 'mcnp6' if code == 'mcnp' else 'sss2'
+        if not shutil.which(executable, os.X_OK):
```
Comment (paulromano) [inline neutron-physics/model.py]: Totally forgot that Python has `which`! I love it

File: neutron-physics/model.py [hunk 4]
```diff
@@ -200,7 +342,7 @@ def _make_openmc_input(self):
         materials.export_to_xml(Path('openmc') / 'materials.xml')
 
         # Set up geometry
-        sphere = openmc.Sphere(boundary_type='reflective', r=1.e9)
+        sphere = openmc.Sphere(boundary_type='vacuum', r=1.e9)
```
Comment (paulromano) [inline neutron-physics/model.py]: Is this intentional?

File: neutron-physics/model.py [hunk 5]
```diff
@@ -240,44 +383,47 @@ def _make_mcnp_input(self):
...
         lines.append('2 0 1 imp:n=0')
+        lines.append('')
  
         # Create the surface cards: sphere centered on origin with 1e9 cm
-        # radius and reflective boundary conditions
-        lines.append('')
+        # radius and vacuum boundary conditions
         lines.append('c --- Surface cards ---')
-        lines.append('*1 so 1.0e9')
+        lines.append('+1 so 1.e9')
```
Comment (paulromano) [inline neutron-physics/model.py]: Ok, now I'm more confused. I had to look this up to figure out it is a white boundary condition. What is the intent of this change?

File: neutron-physics/validate.py [hunk 6]
```diff
@@ -14,15 +14,22 @@
...
-parser.add_argument('-s', '--suffix', default='70c',
+parser.add_argument('-c', '--code', choices=['mcnp', 'serpent'], default='mcnp',
+                    help='Code to validate OpenMC against.')
+parser.add_argument('-s', '--suffix', type=str, default='70c',
                     help='MCNP cross section suffix')
-parser.add_argument('-l', '--library', type=str,
-                    help='Directory containing endf70[a-k] or endf71x MCNP ACE '
-                    'data library. If specified, an HDF5 library that can be '
-                    'used by OpenMC will be created from the MCNP data.')
+parser.add_argument('-l', '--library', type=str, help='XSDIR directory file. '
```
Comment (paulromano) [inline neutron-physics/validate.py]: Maybe we should make this `-x` / `--xsdir` to be more explicit?

File: photon-physics/model.py [hunk 7]
```diff
@@ -54,61 +62,146 @@ class Model(object):
...
+    def particles(self, particles):
+        if particles % self._batches != 0:
+            msg = (f'Number of particles {particles} must be divisible by '
+                   f'the number of batches {self._batches}.')
+            raise ValueError(msg)
+        self._particles = particles
+
+    @code.setter
+    def code(self, code):
+        if code not in ['mcnp', 'serpent']:
```
Comment (paulromano) [inline photon-physics/model.py]: use tuple instead of list

File: photon-physics/model.py [hunk 8]
```diff
@@ -54,61 +62,146 @@ class Model(object):
...
+            msg = (f'Unsupported code {code}: code must be either "mcnp" or '
+                   f'"serpent".')
```
Comment (paulromano) [inline photon-physics/model.py]: No f-prefix needed on second literal

File: photon-physics/model.py [hunk 9]
```diff
@@ -54,61 +62,146 @@ class Model(object):
...
+            raise ValueError(msg)
+        executable = 'mcnp6' if code == 'mcnp' else 'sss2'
+        if not shutil.which(executable, os.X_OK):
+            msg = f'Unable to locate executable {executable} in path.'
+            raise ValueError(msg)
+        self._code = code
+
+    @suffix.setter
+    def suffix(self, suffix):
+        if not re.match('0[1-4]p|12p', suffix):
```
Comment (paulromano) [inline photon-physics/model.py]: Isn't there an 84p too?

File: photon-physics/model.py [hunk 10]
```diff
@@ -232,36 +401,160 @@ def _make_mcnp_input(self):
...
+        # Turn on unresolved resonance probability treatment
+        lines.append('set ures 1')
+
+        # Set electron treatment
+        if self.electron_treatment == 'led':
+            lines.append('set ttb 0')
+        else:
+            lines.append('set ttb 1')
+
+        # Turn on Doppler broadening of Compton scattered photons (on by
```
Comment (paulromano) [inline photon-physics/model.py]: Should that be "off by default"?

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for the PR! I like how this is set up where you reference the xsdir and then automatically create the xsdata for Serpent. I am noticing quite a bit of duplicated code across our three cases -- we should think about how to reduce some of the redundancy there.

**Comment by amandalund** [comment]
Sorry for the delay! I addressed those comments and did some reorganizing to get rid of a bunch of the redundant code.

File: setup.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env python
+
+from setuptools import setup
+
+
+setup(
+    name='validation',
+    author='The OpenMC Development Team',
+    author_email='openmc-dev@googlegroups.com',
+    description=('A collection of validation scripts for OpenMC'),
```
Comment (paulromano) [inline setup.py]: Parentheses are superfluous

File: README.md [hunk 12]
```diff
@@ -2,5 +2,6 @@
...
+- **photon-physics** -- A set of scripts for validating photon transport in OpenMC. The scripts generate an infinite medium problem of a single element with a point source at a single energy and compare the resulting photon energy spectra from OpenMC and either MCNP6 or Serpent 2. These scripts require that you have MCNP6 or Serpent 2 installed (`mcnp6` or `sss2` executable available).
+- **photon-production** -- A set of scripts for validating photon production in OpenMC. The scripts generate a broomstick problem where monoenergetic neutrons are shot down a thin, infinitely long cylinder. The energy spectra of photons produced from neutron reactions is tallied along the surface of the cylinder. Comparisons are made between OpenMC and either MCNP6 or Serpent 2. Again, these scripts require that MCNP6 or Serpent 2 is properly installed.
```
Comment (paulromano) [inline README.md]: README needs updating since these directories no longer exist

File: setup.py [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+    url='https://github.com/openmc-dev/validation',
+    packages=['validation'],
+    entry_points={
+        'console_scripts': [
+            'validate-neutron-physics=validation.neutron_physics:main',
+            'validate-photon-physics=validation.photon_physics:main',
+            'validate-photon-production=validation.photon_production:main'
```
Comment (paulromano) [inline setup.py]: Maybe we should prefix these scripts with `openmc-` to make it clear what they are related to

File: validation/neutron_physics.py [hunk 14]
```diff
@@ -0,0 +1,532 @@
...
+    args = parser.parse_args()
+
+    model = NeutronPhysicsModel(
+        args.nuclide, args.density, args.energy, args.particles, args.code,
+        args.suffix, args.xsdir, args.thermal, args.output_name
+    )
+    model.run()
+
+
+class NeutronPhysicsModel(object):
```
Comment (paulromano) [inline validation/neutron_physics.py]: `class NeutronPhysicsModel:` will suffice

**Review by paulromano — APPROVED** [review]

**Pull request merged**